entity-framework/core/querying/user-defined-function-mapping.md

```diff
@@ -0,0 +1,154 @@
+
+And the following model configuration:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#EntityConfiguration)]
+
+Blog can have many posts, each post can be tagged with multiple tags, and each tag can be associated with multiple posts; we have set up a many-to-many relationship.
+
+Next, create the user-defined function `CommentedPostCountForBlog`, which returns the count of posts with at least one comment for a given blog, based on the blog `Id`:
+
+```sql
+CREATE FUNCTION dbo.CommentedPostCountForBlog(@id int)
+RETURNS int
+AS
+BEGIN
+    RETURN (SELECT COUNT(*)
+        FROM [Posts] AS [p]
+        WHERE ([p].[BlogId] = @id) AND ((
+            SELECT COUNT(*)
+            FROM [Comments] AS [c]
+            WHERE [p].[PostId] = [c].[PostId]) > 0));
+END
+```
+
+To use this function in EF Core, we define the following CLR method, which we map to the user-defined function:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#BasicFunctionDefinition)]
+
+In the example, the method is defined on `DbContext`, but it can also be defined as a static method in other places. Note that the body of the CLR method is not important, as it will never be invoked client-side; EF Core only looks at the method signature.
+
+This function definition can now be associated with user-defined function in the model configuration:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#BasicFunctionConfiguration)]
+
+> [!NOTE]
+> By default, EF Core tries to map CLR function to a user-defined function with the same name. If the names differ, we can use `HasName` to provide the correct name for the user-defined function we want to map to.
+
+Now, executing the following query:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Program.cs#BasicQuery)]
+
+Will produce this SQL:
+
+```sql
+SELECT [b].[BlogId], [b].[Rating], [b].[Url]
+FROM [Blogs] AS [b]
+WHERE [dbo].[CommentedPostCountForBlog]([b].[BlogId]) > 1
+```
+
+## Mapping a method to a custom SQL
+
+EF Core also allows for user-defined functions that get converted to a specific SQL. This can be done by specifying a SQL expression using the [Microsoft.EntityFrameworkCore.Query.SqlExpressions](/dotnet/api/microsoft.entityframeworkcore.query.sqlexpressions) API. The SQL expression is provided using `HasTranslation` method during user-defined function configuration.
+
+In the example below, we'll create a function that computes difference between two integers.
+
+The CLR method is as follows:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#HasTranslationFunctionDefinition)]
+
+The function definition is as follows:
+
+[!code-csharp[Main](../../../samples/core/Querying/UserDefinedFunctionMapping/Model.cs#HasTranslationFunctionConfiguration)]
```

Comment:
Avoid using GetService inside OnModelCreating. Use barebone new SqlExpression instead.
Perhaps translate it as Abs(a - b) so you can flow typemapping directly.

Reply:
i tried that in the first iteration and it didn't look great - SqlFunctionExpression ctor is quite complicated also creating a function call inside a mapped function is bit weird, will try to do the case and see if its any better. Whats the problem with using get service here?

Reply:
Starts wrong guidance and then people will start calling GetService with a lot of random other services. Not all of that work. We have a tracking issue to make experience a better but till we have that, we should just guide people to barebone new SqlExpression syntax.

Reply:
sounds good, ended up constructing necessary type mappings manually. not the prettiest but gets the job done ;)